__experimentalBlockPatterns: load them with REST API

jsnajdr opened the pull request:
This is a draft fix for #39055. We're removing `__experimentalBlockPatterns` data from editor settings, stop sending them in the initial HTML document, and load them from a new REST endpoint instead. The PR has several parts:

1. Add a new `/wp/v2/block-patterns` REST endpoint. Currently, it only supports getting the whole list, not individual items. Gutenberg doesn't need the individual items, but I guess it's a good custom in WP REST APIs to provide access to them, too.
2. Add a new `getBlockPatterns()` selector/resolver to the `core-data` store. It uses the new endpoint to fetch the data.
3. In the Post Editor React UI, immediately before the `settings` object is passed to `BlockEditorProvider`, fill the `__experimentalBlockPatterns` field with results of the `core-data` selector if it's not present.
4. Do the same also in the Site Editor UI, at a different place, but in the same spirit. We can also remove the creation of `__experimentalBlockPatterns` field in the PHP code that generates the `wp.editSite.initializeEditor` code.
5. Register a filter (`block_editor_settings_all`) that removes `__experimentalBlockPatterns` field from the settings generated by Core pages like `site-editor.php` or `edit-form-blocks.php`.

If this works well, I'll add an endpoint for block pattern categories, too.

There's one thing I don't like about this patch: I used an approach that is already used for `__experimentalReusableBlocks`, but it has one downside: instead of loading the block patterns only when they're needed, i.e., when opening the Block Inserter UI, they are loaded when the editor is initialized. So the download is not really delayed, only slightly deferred and parallelized: instead of being a static part of the initial HTML, there's a separate REST request that doesn't block anything.

I'll proceed with trying to improve this: the `settings` field won't be the block pattern list itself, but rather a function to fetch them, used by a resolver. There's an `__experimentalFetchLinkSuggestions` field that works like this.

jsnajdr committed:
Add REST endpoint to fetch registered block patterns

jsnajdr committed:
Core Data: add getBlockPatterns selector (with resolver and state)

jsnajdr committed:
Editor: fetch block patterns from REST if not supplied in editor settings

jsnajdr committed:
Site Editor: fetch block patterns from REST, don't preload in settings

jsnajdr committed:
Remove __experimentalBlockPatterns from settings generated by Core

jsnajdr committed:
Re-add the request param

ntsekouras commented:
>There's one thing I don't like about this patch: I used an approach that is already used for __experimentalReusableBlocks, but it has one downside: instead of loading the block patterns only when they're needed, i.e., when opening the Block Inserter UI, they are loaded when the editor is initialized.

I think that's good and is needed because patterns are not only used in the inserter. They are used in every block in block-switcher trying to find pattern transformations, during some blocks insertion, etc.... Because parsing the patterns can take some time, right now we preparse them on load, on browsers idle time and I think this should be preserved.

spacedmonkey commented on lib/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,121 @@
+	 */
+	public function get_items() {
+		$data = WP_Block_Patterns_Registry::get_instance()->get_all_registered();
+		return rest_ensure_response( $data );
```
Please ensure that `prepare_item_for_response` is implemented here. I would also use `prepare_response_for_collection`. 

This is a good example to follow. 

https://github.com/WordPress/wordpress-develop/blob/c7e067c578362fd67580ef1ef230856e57171329/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L142-L154

jsnajdr replied:
I implemented `prepare_item_for_response`, which merely copies the supported fields from the "raw" pattern to the "prepared" one. There's not need to sanitize or escape any fields. The patterns have been either registered by themes and plugins, or downloaded from wordpress.org and already sanitized. There's also an `add_additional_fields_to_object` call which allows new fields to be registered with `register_rest_field( 'block-pattern' )`.

I added also a `prepare_response_for_collection` call, although it's a noop because there are no links in the items.

spacedmonkey commented on lib/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,121 @@
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'wp/v2';
```
Testing REST API endpoints, use the __experimental namespace.

jsnajdr replied:
Moved to `__experimental` namespace ✅

spacedmonkey commented on lib/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,121 @@
+ *
+ * @see   WP_REST_Controller
+ */
+class WP_REST_Block_Patterns_Controller extends WP_REST_Controller {
```
All methods in this function need since documentation.

ntsekouras replied:
Also we should move this in `compat/wordpress-6.0` folder.

jsnajdr replied:
All methods of the controller are documented, and I moved the controller to `compat/wordpress-6.0`. The endpoint should be moved to Core for 6.0 release.

spacedmonkey reviewed (changes requested):
When providing this much PHP code, please provide unit test to prove the REST API endpoints work. Without them it hard to review. 

Please ensure that passing `?_fields=` param to the rest api works, as other endpoints do.

jsnajdr commented:
> I think that's good and is needed because patterns are not only used in the inserter.

Still, they should not block the initial load that leads to displaying the post and edit UI. The first time I really need the patterns is when I click on a block and its block controls are displayed, right? Only then there is UI to transform the block to something else, to replace it with another pattern etc.

The canonical way to lazy load things is that they are not loaded until their selector is actually called for the first time:
```js
useSelect( ( select ) => select( blockEditorStore ).getPatterns() );
```
In our case though, the patterns are loaded much earlier, namely when mounting the editor provider component.

I'm not sure how much of a problem this will still be after we fully migrate the load to REST API. The ideal situation though is that they are loaded as late as possible.

jsnajdr committed:
Move the patterns endpoint to experimental namespace, change name

ntsekouras commented on lib/rest-api.php:
```diff
@@ -40,6 +40,15 @@ function gutenberg_register_rest_navigation_areas() {
+	$block_patterns = new WP_REST_Block_Patterns_Controller();
+	$block_patterns->register_routes();
+}
+add_action( 'rest_api_init', 'gutenberg_register_rest_block_patterns' );
```
We should move this in `compat/wordpress-6.0` folder.

ntsekouras commented on lib/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,121 @@
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items() {
```
Without the `$request` param this throws a warning.
```suggestion
	public function get_items( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
```

ntsekouras reviewed (commented):
I tested and this seems to work well. Thanks @jsnajdr!

jsnajdr committed:
Disable lint error about unused parameter

jsnajdr committed:
Sync schema descriptions with pattern directory endpoint

jsnajdr committed:
Move the REST endpoint to lib/compat/wordpress-6.0

jsnajdr committed:
Load remote block patterns when processing the REST request

jsnajdr committed:
Implement prepare_item_for_response

jsnajdr committed:
Prevent loading remote block patterns in site/post editor, limit to REST

jsnajdr commented:
> Please ensure that passing `?_fields=` param to the rest api works, as other endpoints do.

That's supported without writing any explicit code, because the `_fields` are filtered by the `rest_filter_response_fields` function registered with the `rest_post_dispatch` hook.

The endpoint handler could call `$this->get_fields_for_response` and populate only the requested fields, but that's useful only when it's expensive to calculate the fields. Here we're just copying strings from one object to another.

jsnajdr committed:
Fix name of the REST controller file

spacedmonkey commented on lib/compat/wordpress-6.0/class-gutenberg-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,154 @@
+	 * @return WP_REST_Response
+	 */
+	public function prepare_item_for_response( $item, $request ) {
+		$prepared_pattern = array(
+			'name'           => $item['name'],
+			'title'          => $item['title'],
+			'blockTypes'     => $item['blockTypes'],
+			'categories'     => $item['categories'],
+			'content'        => $item['content'],
+		);
+
+		$prepared_pattern = $this->add_additional_fields_to_object( $prepared_pattern, $request );
```
```suggestion
                 $fields     = $this->get_fields_for_response( $request );
		$prepared_pattern = array();
		$keys = array( 'name', 'title', 'blockTypes', 'categories', 'content' );
		foreach( $keys as $key ) {
		    if ( rest_is_field_included( $key, $fields ) ) {
			$prepared_pattern[$key] = $item[$key];
		     }
		}

		$prepared_pattern = $this->add_additional_fields_to_object( $prepared_pattern, $request );
```

Let's try this.

spacedmonkey replied:
@jsnajdr This still needs eyes on.

jsnajdr replied:
The `_fields` list is already being honored, just in a different way, I described it here: https://github.com/WordPress/gutenberg/pull/39185#issuecomment-1070924513

Or is it preferred that `rest_is_field_included` is used explicitly by all new endpoints?

spacedmonkey commented on lib/compat/wordpress-6.0/class-gutenberg-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,154 @@
+
+		$prepared_pattern = $this->add_additional_fields_to_object( $prepared_pattern, $request );
+
+		return new WP_REST_Response( $prepared_pattern );
```
```suggestion
		$context = ! empty( $request['context'] ) ? $request['context'] : 'view';
		$data    = $this->add_additional_fields_to_object( $prepared_pattern, $request );
		$data    = $this->filter_response_by_context( $data, $context );

		$response = rest_ensure_response( $data );
		
		return $response;
```

jsnajdr replied:
Added context support to the endpoint, although in this case it doesn't really make much difference.

jsnajdr committed:
Fix code style nits reported by linter

spacedmonkey commented on lib/compat/wordpress-6.0/class-gutenberg-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,154 @@
+				'title'      => array(
+					'description' => __( 'The pattern title, in human readable format.', 'gutenberg' ),
+					'type'        => 'string',
+					'readonly'    => true,
```
Please ensure that all these properties have `'context' ` defined.

jsnajdr committed:
Add context support

jsnajdr committed:
Stop calling gutenberg_register_remote_theme_patterns in site/post editor

jsnajdr committed:
Document filter param

jsnajdr commented:
@ntsekouras There are some changes in this PR that are related to #39493 and that you might be the best person to review:

the goal is to load the remote patterns only in the `block-patterns/patterns` REST endpoint and to disable loading them everywhere else, with the `should_load_remote_block_patterns` filter. To achieve that, the PR does:
- update the `lib/compat/wordpress-5.9/block-patterns.php` file to only create the `_load_remote_featured_patterns` function, which otherwise doesn't exist in 5.9 (only in 6.0), and stop calling it in the `current_screen` action -- we don't want to call it anywhere outside the REST endpoint.
- update the `lib/compat/wordpress-6.0/block-patterns.php` file to only create the `gutenberg_register_remote_theme_patterns` function (new in 6.0) and stop calling it in the `current_screen` action. Various checks (is theme support, is disabled, ...) are move from the action to the function itself.
- in `lib/compat/wordpress-6.0/edit-form-blocks.php`, add a new action that disables remote patterns (with `should_load_remote_block_patterns`) for block editor pages -- that's to neutralize existing Core calls to `_load_remote_featured_patterns` et al.
- load all the patterns (core, featured, theme) explicitly in the REST controller for `block-patterns/patterns`.

jsnajdr committed:
Add REST endpoint for block pattern categories

jsnajdr committed:
Core Data: add getBlockPatternCategories selector (with resolver and state)

jsnajdr committed:
Load block pattern categories from REST instead of settings

jsnajdr committed:
Remove block pattern categories from server-generated editor settings

jsnajdr committed:
Fix spacing issues

jsnajdr commented:
@spacedmonkey @ntsekouras I believe I addressed all review feedback and that the PR is ready for reviewing and merging. The major changes the PR does are:
1. Introduce two new REST endpoints, `__experimental/block-patterns/patterns` and `.../categories`, which return the registered patterns and categories. The REST callback does the actual registration (the sources are wordpress.org and current theme). All the formal rules for WP REST endpoints (schemas, contexts, etc.) should be followed by now.
2. Introduce selectors to the `core-data` store for fetching from these two endpoints.
3. Patch the `block-editor` providers for post and site editors to load patterns and categories from `core-data` instead from settings.
4. Stop adding `__experimentalBlockPatterns` and `__experimentalBlockPatternCategories` to the block editor settings object serialized in the initial HTML.
5. Prevent registering block patterns on all code paths except the REST endpoint. That's now the only place where we need to register them.

jsnajdr commented:
Ah, I still need to add unit tests for the REST endpoints and the new data entities, too. That's the only missing part I'm aware of.

spacedmonkey commented:
@jsnajdr  This still needs to be actioned. https://github.com/WordPress/gutenberg/pull/39185#discussion_r829117790

We also PHP unit tests. But we are nearly there.

jsnajdr committed:
Return only requested fields

jsnajdr committed:
Unit tests for the REST controllers

jsnajdr commented:
@spacedmonkey I added `get_fields_for_response` processing to both endpoints, and added a unit test for both of them. Please have a look and let me know if there's something still missing, e.g., something in the tests that's definitely worth testing and isn't.

ntsekouras commented on lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,161 @@
+	 */
+	public function prepare_item_for_response( $item, $request ) {
+		$fields = $this->get_fields_for_response( $request );
+		$keys   = array( 'name', 'title', 'blockTypes', 'categories', 'content' );
```
We should add the [remaining fields ](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/) here.

ntsekouras commented on packages/core-data/src/resolvers.js:
```diff
@@ -453,3 +453,17 @@ export const __experimentalGetCurrentThemeGlobalStylesVariations = () => async (
+};
+
+export const getBlockPatternCategories = () => async ( { dispatch } ) => {
+	const patterns = await apiFetch( {
+		path: '/__experimental/block-patterns/categories',
+	} );
+	dispatch( { type: 'RECEIVE_BLOCK_PATTERN_CATEGORIES', patterns } );
```
```suggestion
	const categories = await apiFetch( {
		path: '/__experimental/block-patterns/categories',
	} );
	dispatch( { type: 'RECEIVE_BLOCK_PATTERN_CATEGORIES', categories } );
```

jsnajdr replied:
Done :+1:

ntsekouras commented on phpunit/class-wp-rest-block-patterns-controller-test.php:
```diff
@@ -0,0 +1,65 @@
+
+	public function test_get_items() {
+		// Change this when the registered Core patterns change.
+		$expected_names  = array(
```
Should we make this more resilient by probably unregistering existing patterns, register some test patterns and then restore?

jsnajdr replied:
Yes, I'll try to do that, the tests are too fragile in the current form. I'll probably have to create a mock version of the two registries.

ntsekouras commented on lib/compat/wordpress-6.0/edit-form-blocks.php:
```diff
@@ -48,5 +48,19 @@ function optimize_preload_paths( $preload_paths ) {
+ */
+function disable_load_remote_patterns( $current_screen ) {
+	$is_site_editor = ( function_exists( 'gutenberg_is_edit_site_page' ) && gutenberg_is_edit_site_page( $current_screen->id ) );
+	if ( $is_site_editor || $current_screen->is_block_editor() ) {
```
I'm a bit confused here. Why do we need to check for the current screen right now and not just add the filter always? And by
>Disables loading remote block patterns from REST while initializing the editor

you mean the existing call to pattern directory, correct?

jsnajdr replied:
> Why do we need to check for the current screen right now and not just add the filter always?

I want to disable the existing remote loading calls that are performed in Core, on the post and site editor pages. That's why I'm targeting the filter to these pages.

> you mean the existing call to pattern directory, correct?

Yes, this is solely to disable existing calls in Core. When this compat code is merged to 6.0, these calls will be removed and this `disable_load_remote_patterns` function will disappear.

ntsekouras reviewed (commented):
Thanks @jsnajdr!

ntsekouras commented:
We have one more issue in this line: https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-5.9/block-patterns.php#L26. When me make the request to get the `pattern categories` the `featured` category hasn't been registered yet, because this happens to the request to `patterns`.

jsnajdr committed:
Fixup action field name when receiving categories

jsnajdr committed:
Use mock registries in unit tests

jsnajdr committed:
Add missing pattern properties

jsnajdr commented:
@ntsekouras I addressed all pending review issues:
- in unit tests, the pattern and category registries are now mocked, using an empty instances where only testing items are registered. The tests are isolated from the actual patterns and categories that Core registers.
- I added support for the `description`, `keywords` and `viewportWidth` pattern fields. I originally thought they are legacy fields that are no longer used, but that was wrong. They all have an active role.

jsnajdr commented:
There are e2e tests that are repeatedly failing, in `specs/site-editor/style-variations.test.js`. Let's figure out if that's my fault or not.

spacedmonkey commented on lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,188 @@
+			}
+		}
+
+		$context = ! empty( $request['context'] ) ? $request['context'] : 'view';
```
Just notice we have categories are block types here. We could add some links to block types api and the block category. Even allow them to be embedded.

spacedmonkey commented on lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,188 @@
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		// Load block patterns from w.org.
```
Is there value in allow to limit the query by category?

spacedmonkey commented on lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,188 @@
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+				),
+				'schema'      => array( $this, 'get_public_item_schema' ),
+				'allow_batch' => array( 'v1' => true ),
```
Why is batching support needed here?

spacedmonkey commented on lib/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php:
```diff
@@ -0,0 +1,188 @@
+	 * @see   register_rest_route()
+	 */
+	public function register_routes() {
+		register_rest_route(
```
Is there anyway to get a single block pattern? Like `__experimental/block-patterns/patterns/<id>`

spacedmonkey commented on lib/compat/wordpress-6.0/class-wp-rest-block-pattern-categories-controller.php:
```diff
@@ -0,0 +1,138 @@
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+				),
+				'schema'      => array( $this, 'get_public_item_schema' ),
+				'allow_batch' => array( 'v1' => true ),
```
Why is batching support needed here.

spacedmonkey commented on lib/compat/wordpress-6.0/class-wp-rest-block-pattern-categories-controller.php:
```diff
@@ -0,0 +1,138 @@
+	 * @see   register_rest_route()
+	 */
+	public function register_routes() {
+		register_rest_route(
```
Is there anyway to get a single block pattern categories? Like __experimental/block-patterns/categories/<id>

spacedmonkey commented on phpunit/class-wp-rest-block-pattern-categories-controller-test.php:
```diff
@@ -0,0 +1,68 @@
+<?php
+
```
PHP docs would be good here.

spacedmonkey commented on phpunit/class-wp-rest-block-patterns-controller-test.php:
```diff
@@ -0,0 +1,83 @@
+<?php
+
```
PHP docs would be good here.

jsnajdr committed:
Restore original registry instance after test is finished

jsnajdr committed:
Add PHP docs comments to unit test suites

jsnajdr commented:
@spacedmonkey I added a PHP docblock to both unit tests. I documented only the test class itself, not the individual methods. There's nothing to be said about standard methods like `set_up`), and individual test cases typically have doc comments only when they are associated with particular ticket, which is not the case here.

About other suggestions:

> Why is batching support needed here?

There's no reason why the request couldn't be a part of a batch, so I'm enabling batching support. Because why not? My understanding is that it's disabled by default.

> Just notice we have categories are block types here. We could add some links to block types api and the block category. Even allow them to be embedded.

To make the `blockTypes` array embeddable, I would first need to add a `"_links": { "blockTypes": "https://..." }` field, but what would the link point to? When a post, for example, has two categories `categories: [ 1, 2 ]`, the link to them is like:
```
/categories?post=588
```
and it returns an array of the two categories objects, with IDs `1` and `2`. How to embed `blockTypes`? A link like:
```
/block-types?block-pattern=core/query-offset-posts
```
would have to work, but it doesn't. The `/block-types` endpoint doesn't support such a filter.

Supporting links and embeds, filtering by category, and fetching individual items, they all look like interesting things to have, but there's currently no use case for them -- the editor only fetches the full list. That's why I think they are out of scope for this PR, which is already quite big. I'd prefer to work on them separately, after this PR ships.

spacedmonkey reviewed (approved):
Approve, all other tweaks I have suggested, can happen in follow up PRs.

spacedmonkey commented:
> would have to work, but it doesn't. The /block-types endpoint doesn't support such a filter.

What I mean was this. `http://localhost:8899/wp-json/wp/v2/block-types/core/block` which the block type support rest api does support.

jsnajdr commented:
> `http://localhost:8899/wp-json/wp/v2/block-types/core/block`

That would work if the block pattern was linking to just one block type. Then it's similar to how, e.g., `author` can be linked and embedded:
```
{
  "blockType": "core/paragraph",
  "_links": {
    "blockType": [ {
      "embeddable": true,
      "href": "http://localhost:8899/wp-json/wp/v2/block-types/core/paragraph"
    } ],
  },
  "_embedded: {
    "blockType": [ {
      "name": "core/paragraph",
      ...
    } ]
  }
}
```
How would the structure look like for array of `blockTypes`? Would the `_links.blockTypes` field be an array of several links to individual block types?

ntsekouras reviewed (approved):
Thanks @jsnajdr! LGTM 👍 

I seem to cannot reproduce my previous [comment here](https://github.com/WordPress/gutenberg/pull/39185#issuecomment-1076352895) about the `featured` category not being fetched and I couldn't see any related changes about it, afterwards, so 🤷 ..

TimothyBJacobs commented:
> > Why is batching support needed here?
> There's no reason why the request couldn't be a part of a batch, so I'm enabling batching support. Because why not? My understanding is that it's disabled by default.

Batching is not possible at all for `GET` requests. So we should remove this in a follow up PR.

jsnajdr commented:
> Batching is not possible at all for `GET` requests. So we should remove this in a follow up PR.

That's right, fixing in #39834.
